core/Segment.php

```diff
@@ -460,8 +465,42 @@ public function getHash()
 
     public static function getSegmentHash($definition)
     {
-        // urldecode to normalize the string, as browsers may send slightly different payloads for the same archive
-        return md5(urldecode($definition));
+        $cache = Cache::getEagerCache();
+        $cacheKey = self::CACHE_KEY . md5($definition);
+
+        if ($cache->contains($cacheKey)) {
+            return $cache->fetch($cacheKey);
+        }
+
+        $defaultHash = md5(urldecode($definition));
+
+        // if the cache for segments already built, but this segment was not found,
+        // we return the default segment, this can be a segment from url or
+        // something like "visitorType==new"
+        if ($cache->contains(self::SEGMENT_HAS_BUILT_CACHE_KEY)) {
+            return $defaultHash;
+        }
+
+        // the segment hash is not built yet, let's do it
+        $cache->save(self::SEGMENT_HAS_BUILT_CACHE_KEY, true);
+
+        $model = new SegmentEditorModel();
+        $segments = $model->getAllSegmentsForAllUsers();
+        $foundSegment = false;
+
+        foreach ($segments as $segment) {
+            $cacheKey = self::CACHE_KEY . md5($segment['definition']);
+
+            $cache->save($cacheKey, $segment['hash']);
+
+            if ($segment['definition'] === $definition || $segment['definition'] === urldecode($definition) || $segment['definition'] === urlencode($definition)) {
```

Comment:
not sure I understand this part? I think I would have maybe expected something like this:

```php
foreach ($segments as $segment) {
            $cacheKey = self::CACHE_KEY . md5($segment['definition']);
            $cache->save($cacheKey, $segment['hash']);
            
            $cacheKey = self::CACHE_KEY . md5(urldecode($segment['definition']));
            $cache->save($cacheKey, $segment['hash']);
            
            $cacheKey = self::CACHE_KEY . md5(urlencode($segment['definition']));
            $cache->save($cacheKey, $segment['hash']);
}
   if ($cache->contains($cacheKey)) {
            return $cache->fetch($cacheKey);
        }
        return $defaultHash;
```

Not sure if that makes any sense though? I'm just wondering if otherwise we wouldn't find the correct cache key depending if the segment definition was passed urlencoded or urldecoded etc. Not sure it's clear what I mean?

Reply:
I see what you mean, we save all versions for every cache. But the `cache->contains` call after this will check all the caches, even that is not stored in the file yet, only saved? Is that correct?

Reply:
I'm not 100% sure what you mean but I think it's correct. 👍